internal/sink/sink.go

```diff
@@ -0,0 +1,544 @@
+	for _, message := range messages {
+		offsetStore.Add(message.KafkaMessage.TopicPartition)
+	}
+	offsets := offsetStore.Get()
+	if len(offsets) > 0 {
+		commitedOffsets, err := s.consumer.CommitOffsets(offsets)
+		if err != nil {
+			logger.Error("failed to commit offset to kafka", "err", err)
+			// do not return error here as we want to write Redis even if commit fails as we already saved them in ClickHouse
+		}
+		logger.Debug("succeeded to commit offset to kafka", "offsets", commitedOffsets)
+	}
+
+	// 3. Sink to Redis
+	// Least once guarantee, if Redis write fails we will accept messages with same idempotency key in future and
+	// we have to relay on ClickHouse's deduplication.
+	// Deduplicator is an optional dependency so we check if it's set
+	if s.config.Deduplicator != nil && len(dedupedMessages) > 0 {
+		dedupeItems := []dedupe.Item{}
+		for _, message := range dedupedMessages {
+			dedupeItems = append(dedupeItems, dedupe.Item{
+				Namespace: message.Namespace,
+				ID:        message.Serialized.Id,
+				Source:    message.Serialized.Source,
+			})
+		}
+
+		err := s.config.Deduplicator.Set(s.config.Context, dedupeItems...)
+		if err != nil {
+			logger.Error("failed to sink to redis", "err", err)
+			return fmt.Errorf("failed to sink to redis: %s", err)
+		}
+		logger.Debug("succeeded to sink to redis", "buffer size", len(messages))
+	}
+
+	if len(messages) > 0 {
+		logger.Debug("succeeded to flush", "buffer size", len(messages))
+	}
+
+	return nil
+}
+
+// deadLetter sends a message to the dead letter queue, useful permanent non-recoverable errors like json parsing
+func (s *Sink) deadLetter(messages ...SinkMessage) error {
+	logger := s.config.Logger.With("sink", "deadLetter")
+
+	for _, message := range messages {
+		topic := fmt.Sprintf(s.config.DeadletterTopicTemplate, message.Namespace)
+		headers := message.KafkaMessage.Headers
+		headers = append(headers, kafka.Header{Key: "error", Value: []byte(message.Error.Error())})
+
+		msg := &kafka.Message{
+			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+			Timestamp:      message.KafkaMessage.Timestamp,
+			Headers:        headers,
+			Key:            message.KafkaMessage.Key,
+			Value:          message.KafkaMessage.Value,
+		}
+
+		err := s.producer.Produce(msg, nil)
+		if err != nil {
+			return fmt.Errorf("producing kafka message to deadletter topic: %w", err)
+		}
+	}
+
+	logger.Debug("succeeded to deadletter", "messages", len(messages))
+	return nil
+}
+
+func (s *Sink) getNamespaces() (*NamespaceStore, error) {
+	meters, err := s.config.Storage.GetMeters(s.config.Context)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get meters: %s", err)
+	}
+
+	namespaceStore := NewNamespaceStore()
+	for _, meter := range meters {
+		namespaceStore.AddMeter(meter.Namespace, meter)
+	}
+
+	return namespaceStore, nil
+}
+
+func (s *Sink) subscribeToNamespaces() error {
+	logger := s.config.Logger.With("sink", "subscribeToNamespaces")
```

Comment:
```suggestion
	logger := s.config.Logger.With("operation", "subscribeToNamespaces")
```

We could add a "component => sink" in a higher layer.